source/convection_diffusion_solver.cc

```diff
@@ -0,0 +1,410 @@
+
+    const std::string str_convective_term_time_discretization(prm.get("Convective term time discretization"));
+
+    if (str_convective_term_time_discretization == std::string("semi-implicit"))
+      convective_term_time_discretization
+      = RunTimeParameters::ConvectiveTermTimeDiscretization::semi_implicit;
+    else if (str_convective_term_time_discretization == std::string("explicit"))
+      convective_term_time_discretization
+      = RunTimeParameters::ConvectiveTermTimeDiscretization::fully_explicit;
+    else
+      AssertThrow(false,
+                  ExcMessage("Unexpected identifier for the time discretization "
+                             "of the convective term."));
+
+    preconditioner_update_frequency = prm.get_integer("Preconditioner update frequency");
+    AssertThrow(preconditioner_update_frequency > 0,
+           ExcLowerRange(preconditioner_update_frequency, 0));
+
+    verbose = prm.get_bool("Verbose");
+
+    prm.enter_subsection("Linear solver parameters");
+    {
+      solver_parameters.parse_parameters(prm);
+    }
+    prm.leave_subsection();
+  }
+  prm.leave_subsection();
+}
+
+
+
+template<typename Stream>
+Stream& operator<<(Stream &stream, const ConvectionDiffusionParameters &prm)
+{
+  using namespace RunTimeParameters::internal;
+
+  add_header(stream);
+  add_line(stream, "Convection-diffusion solver parameters");
+  add_header(stream);
+
+  switch (prm.convective_term_weak_form) {
+    case RunTimeParameters::ConvectiveTermWeakForm::standard:
+      add_line(stream, "Convective term weak form", "standard");
+      break;
+    case RunTimeParameters::ConvectiveTermWeakForm::rotational:
+      add_line(stream, "Convective term weak form", "rotational");
+      break;
+    case RunTimeParameters::ConvectiveTermWeakForm::divergence:
+      add_line(stream, "Convective term weak form", "divergence");
+      break;
+    case RunTimeParameters::ConvectiveTermWeakForm::skewsymmetric:
+      add_line(stream, "Convective term weak form", "skew-symmetric");
+      break;
+    default:
+      AssertThrow(false, ExcMessage("Unexpected type identifier for the "
+                                    "weak form of the convective term."));
+      break;
+  }
+
+  switch (prm.convective_term_time_discretization) {
+    case RunTimeParameters::ConvectiveTermTimeDiscretization::semi_implicit:
+      add_line(stream, "Convective temporal form", "semi-implicit");
+      break;
+    case RunTimeParameters::ConvectiveTermTimeDiscretization::fully_explicit:
+      add_line(stream, "Convective temporal form", "explicit");
+      break;
+    default:
+      AssertThrow(false, ExcMessage("Unexpected type identifier for the "
+                                    "time discretization of the convective term."));
+      break;
+  }
+
+  add_line(stream,
+           "Preconditioner update frequency",
+           prm.preconditioner_update_frequency);
+
+  stream << prm.solver_parameters;
+
+  return (stream);
+}
+
+
+
```

Comment:
Beginning of the modified content of the original file `source/convection_diffusion.cc`.